homeassistant/components/heos/__init__.py

```diff
@@ -219,6 +233,147 @@ def update_ids(self, mapped_ids: dict[int, int]):
+        entity_id_to_player_id_map = self._get_entity_id_to_player_id_map()
+        leader_id = entity_id_to_player_id_map.get(leader_entity_id)
+        if not leader_id:
+            raise HomeAssistantError(
+                f"The group leader {leader_entity_id} could not be resolved to a HEOS player."
+            )
+        member_ids = [
+            entity_id_to_player_id_map[member]
+            for member in member_entity_ids
+            if member in entity_id_to_player_id_map
+        ]
+
+        try:
+            await self.controller.create_group(leader_id, member_ids)
+        except HeosError as err:
+            _LOGGER.error(
+                "Failed to group %s with %s: %s",
+                leader_entity_id,
+                member_entity_ids,
+                err,
+            )
+
+    async def async_unjoin_player(self, player_entity_id: str):
+        """Remove `player_entity_id` from any group."""
+        player_id = self._get_entity_id_to_player_id_map().get(player_entity_id)
+        if not player_id:
+            raise HomeAssistantError(
+                f"The player {player_entity_id} could not be resolved to a HEOS player."
+            )
+
+        try:
+            await self.controller.create_group(player_id, [])
+        except HeosError as err:
+            _LOGGER.error(
+                "Failed to ungroup %s: %s",
+                player_entity_id,
+                err,
+            )
+
+    async def async_update_groups(self, event, data=None):
+        """Update the group membership from the controller."""
+        if event in (
+            heos_const.EVENT_GROUPS_CHANGED,
+            heos_const.EVENT_CONNECTED,
+            SIGNAL_HEOS_PLAYER_ADDED,
+        ):
+            groups = await self.async_get_group_membership()
+            if groups:
+                self._group_membership = groups
+                _LOGGER.debug("Groups updated due to change event")
+                # Let players know to update
+                async_dispatcher_send(self._hass, SIGNAL_HEOS_UPDATED)
+            else:
+                _LOGGER.debug("Groups empty")
+
+    def connect_update(self):
+        """Connect listener for when groups change and signal player update."""
+        self.controller.dispatcher.connect(
+            heos_const.SIGNAL_CONTROLLER_EVENT, self.async_update_groups
+        )
+        self.controller.dispatcher.connect(
+            heos_const.SIGNAL_HEOS_EVENT, self.async_update_groups
+        )
+
+        # When adding a new HEOS player we need to update the groups.
+        async def _async_handle_player_added():
+            # Avoid calling async_update_groups when `DATA_ENTITY_ID_MAP` has not been
+            # fully populated yet. This may only happen during early startup.
+            if (
+                len(self._hass.data[DOMAIN][MEDIA_PLAYER_DOMAIN])
+                <= len(self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP])
+                and not self._initialized
```

Comment:
Is it only needed to update once per config entry setup?

Reply:
Yes, this is just required during initial config entry setup. In all other cases the players signal updated groups through the HEOS-native event `heos_const.EVENT_CONNECTED`.

And in case the config entry is reloaded a new `GroupManager` instance will be created which will be initialized the very same way.